web/docs/operators/transformations/yara.md

```diff
@@ -0,0 +1,147 @@
+# yara
+
+Executes Yara rules on byte streams.
+
+## Synopsis
+
+```
+yara [-C|--compiled-rules] [-f|--fast-scan] <rule> [<rule>..]
+```
+
+## Description
+
+The `yara` operator applies [Yara](https://virustotal.github.io/yara/) rules to
+an input of bytes, emitting rule context upon a match.
+
+We modeled the operator after the official [`yara` command-line
+utility](https://yara.readthedocs.io/en/stable/commandline.html) to enable a
+familiar experience for the command users. Similar to the official `yara`
+command, the operator compiles the rules by default, unless you provide the
+option `-C,--compiled-rules`. To quote from the above link:
+
+> This is a security measure to prevent users from inadvertently using compiled
+> rules coming from a third-party. Using compiled rules from untrusted sources
+> can lead to the execution of malicious code in your computer.
+
+### `-C|--compiled-rules`
+
+Interpret the rules as compiled.
+
+When providing this flag, you must exactly provide one rule path as positional
+argument.
+
+### `-f|--fast-scan`
+
+Enable fast matching mode.
+
+### `<rule>`
+
+The path to the Yara rule(s).
+
+If the path is a directory, the operator attempts to recursively add all
+contained files as Yara rules.
+
+## Examples
+
+Consider this example rule:
+
+```yara
+rule test {
+  meta:
+    string = "string meta data"
+    integer = 42
+    boolean = true
+
+  strings:
+    $foo = "foo"
+    $bar = "bar"
+
+  condition:
+    $foo and $bar
+}
+```
+
+You can produce some test matches easily on the command line by feeding bytes
+into the `yara` operator:
+
+```bash
+echo 'foo barbar baz' |
+  tenzir 'load stdin | yara /tmp/test.yara'
+```
+
+The resulting `yara.match` events look as follows:
+
+```json
+{
+  "rule": {
+    "identifier": "test",
+    "namespace": "default",
+    "string": "foo",
+    "tags": [],
+    "meta": [
+      {
+        "key": "string",
+        "value": "\"string meta data\""
```

Comment:
@jachris I think I ran into something in the series builder when it promotes a type to a common type. The `value` field has a string with inner double quotes that—I think—should not be there. A verified that the original data doesn't have them by trimming the string, but that actually removed characters from the content.

Then I looked into `conflict_builder` and found this TODO, which seems to be the reason for this issue:

```cpp
      // TODO: We chose the strategy of always serializing conflicting types as
      // JSON strings. This means that we can always get the original data by
      // parsing the resulting strings as JSON. However, we could also use
      // different strategies here. For example, we could directly add string
      // values to the builder without JSON, such that they are preserved
      // without extra quotes.
```

Defaulting to JSON is safe, but as a user, I'm not sure if I want that. Of course, we all know that sum types are the proper solution, so I'm fine with waiting ultimately. But intuitively, it's a lot harder to *use* this value if I need strip double quotes, as opposed to, say, working with a boolean `"true"` string. Mostly because it's hard to understand that this inner value is in fact JSON.

Have you considered a simple `to_string` here, trading preservation with usability?

Reply:
We can make this behavior configurable, but I first want to see whether you'd perhaps prefer one of two other solutions:

1. Use `to_string()` before giving the metadata to the series builder. You will not get double-quoting (because there is no type conflict). Your proposed behavior and this solution only differ in the case where there is no type conflict. I would argue that the consistency of always outputting a string is better than occasionally having the correct type here, as adding metadata to an existing yara rule can otherwise change the semantics of existing queries (due to `false` -> `"false`", which would do not have the same behavior).
2. Serialize the metadata as a record instead of a list. This way, you can fully preserve types and its easier to address the individual contents.

Reply:
Both options you listed strike me as superior to what I have started with naively. The record is the cleanest, at a slight cost of schema churn. Whether that's an actual problem we still have to learn, so I'm leaning towards that route.

Reply:
Implemented in [18b684f](https://github.com/tenzir/tenzir/pull/3594/commits/18b684f570467dec9c4fde34d17f263b67e25531).